godotproject/Player.gd

```diff
@@ -135,43 +240,84 @@ func do_movement(delta):
+		handle_normal_horizontal_movement(delta)
 		
 		# Adjust Y velocity based on gravity and jump
 		velocity.y += gravity * delta
+
+		# If we're still in the shorthop section of our jump
 		if jump_shorthop_countdown > 0:
 			jump_shorthop_countdown -= delta
-			if jump_shorthop_countdown <= 0 and not Input.is_action_pressed("jump"):
+			# If we're still shorthopping and the user has let go of jump, then abort out of
+			# the main jump by clearing its timer, and use the shorthop speed instead.
+			if jump_shorthop_countdown <= 0 and not jump_held:
 				jump_frame_countdown = 0
-				velocity.y = jump_speed * shorthop_ratio
+				velocity.y = -shorthop_speed
+		# If we're still jumping, set the user's Y velocity to -jump_speed (-Y is up)
 		if jump_frame_countdown > 0:
 			jump_frame_countdown -= delta
-			velocity.y = jump_speed
-		velocity.y = max(min(velocity.y, 1200), -1200)
+			velocity.y = -jump_speed
+		# Prevent the player from falling faster than their terminal velocity
+		velocity.y = min(velocity.y, terminal_velocity)
+
+		# Actually move the player along the velocity vector.
 		velocity = move_and_slide(velocity, Vector2.UP)
+		# Update our stored floor state.
+		on_floor_last_frame = is_on_floor()
+
 		# Count down the hop sound timer, ensuring it never goes below 0.
 		hop_sound_timer = max(hop_sound_timer - delta, 0)
-		# Play sound effect when timer has expired and frog is walking on floor.
-		if hop_sound_timer <= 0 && is_on_floor() && !is_zero(user_direction.x) && !is_zero(velocity.x):
+		# Play sound effect when timer has expired and frog is walking on floor
+		# (only if user is pointing in the same direction that frog is going)
+		if hop_sound_timer <= 0 && on_floor_last_frame && !is_zero(user_direction.x) && sign(velocity.x) == sign(user_direction.x):
 			# Calculate new timer value, based on period as well as random variance.
 			# This slight variance will hopefully make it sound more natural.
 			# Let P and V be the period and variance, the minimum timer value will be P and the max will be P*(1+V)
 			# In practice, with (P,V) = (0.2,0.1), the min will be 0.2 and the max will be 0.22
 			hop_sound_timer = hop_sound_timer_period * (1 + rand_range(0, hop_sound_timer_period_variance))
 			# Play sound effect.
 			$HopSoundPlayer.play(0)
-	pass
+
+func handle_jump_starting():
+	# Adjust jump start related timers used to permit users to jump a short bit
+	# after they technically left the ground or pressed the jump key.
+	# This is done in many games to prevent the user from feeling like their jump was
+	# "stolen" when they actually pressed the button after leaving the ground.
+	# Also allows them to press the jump button just before hitting the ground and
+	# not have their jump eaten. Technically this allows for the user to buffer a
+	# frame-perfect jump which preserves all their horizontal speed, but we can block
+	# this if users prove to be able to perform these jumps very easily.
```

Comment:
I think this sounds great. I remember hearing the deadcells creator talk about something similar to if you just hit the side corner of a platform (ie not the top) the game would let you land on it still. idk if thats feasible for us but I like the mentality of helping the players out. Make the game difficult in ways that are fun not annoying